gadgets/cmr/config/LandmarkDetection/gadgetron_cmr_landmark_detection_util.py

```diff
@@ -1,28 +1,115 @@
+"""Utility file for cmr landmark detection"""
+
+import os
 import sys
-import torch
-import torch.nn as nn
-import torch.optim as optim
-import torch.nn.functional as F
+import urllib.request
+import hashlib
+import fcntl
 import numpy as np
-import scipy
 from scipy import ndimage
 
-def cpad_2d(data, RO, E1):
-    '''
-    data: [dRO, dE1, N], padded it round center to [RO, E1, N]
-    return value: (s_ro, s_e1), starting point of data in padded array
-    '''
+class model_file_lock:
+    """A python inter-process file lock
+    
+    Adaptive from https://stackoverflow.com/questions/6931342/system-wide-mutex-in-python-on-linux
+    """
+    
+    def __init__(self, model_dest):
+        self.model_dest = model_dest
+    
+    def __enter__ (self):
+        self.fp = open(os.path.join(self.model_dest, "ai_model_lockfile.lck"), 'w')
```

Comment:
The pattern here does not work well if two processes are trying to download two different models in the same folder. They would end up locking each other when they don't need to.

Reply:
By adding the model sha256 to the model path, it should solve the racing condition:

```
        # assemble the source and destination of model
        model_url = model_host + model_file
        destination = os.path.join(model_dest, model_sha256, model_file)
        
        print(f"check_and_get_model,  model_url is {model_url}")
        print(f"check_and_get_model,  destination is {destination}")
        
        with model_file_lock(model_dest):
            # make sure destination exist
            os.makedirs(model_dest, exist_ok=True)
            
            # download the model if not exist
            ...
```

If two processes try to download two version of model, they will have different file locks.

If two processes try to download the same model, one will wait for another to download and release the file lock. Then the second process will continue and find model is already there to be used.